### PR TITLE
Surface dice loss

### PR DESCRIPTION
**Pull request by LorenzLamm**
Added the option to use Surface-Dice as a loss function during training.

Surface-Dice is based on "clDice - a Novel Topology-Preserving Loss Function for Tubular Structure Segmentation"
(https://openaccess.thecvf.com/content/CVPR2021/papers/Shit_clDice_-_A_Novel_Topology-Preserving_Loss_Function_for_Tubular_Structure_CVPR_2021_paper.pdf)

Also fixed some issues for patch extraction (corrected naming), removed wandb tracking (caused dependency issues), fixed bug mentioned in https://github.com/teamtomo/membrain-seg/issues/44, and added printing of training parameter summary.

### CODE REVIEW

File: src/membrain_seg/annotations/extract_patch_cli.py [hunk 1]
```diff
@@ -21,6 +21,11 @@ def extract_patches(
         help="Path to the folder where extracted patches should be stored. \
             (subdirectories will be created)",
     ),
```
Comment (LorenzLamm) [inline src/membrain_seg/annotations/extract_patch_cli.py:23]: Dataset token is now readable as well. This helps to distinguish between different datasets, because we may want to apply different loss functions (particularly Surface-Dice) to some datasets, but not to others.

File: src/membrain_seg/segmentation/cli/train_cli.py [hunk 2]
```diff
@@ -84,6 +87,22 @@ def train_advanced(
             but also severely increases training time.\
                 Pass "True" or "False".',
     ),
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/cli/train_cli.py:89]: surface_dice_tokens: dataset tokens specifying which datasets to apply surface-dice to. 

Needs to be passed as separate arguments:
--surface-dice-tokens ds1 --surface-dice-tokens ds2

I did not find a more elegant way with Typer to pass in a list of strings

File: src/membrain_seg/segmentation/dataloading/memseg_dataset.py [hunk 3]
```diff
@@ -102,6 +101,7 @@ def __getitem__(self, idx: int) -> Dict[str, np.ndarray]:
             "label": np.expand_dims(self.labels[idx], 0),
         }
         idx_dict = self.transforms(idx_dict)
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/dataloading/memseg_dataset.py:103]: dataset token is now returned with every train image

File: src/membrain_seg/segmentation/dataloading/memseg_dataset.py [hunk 4]
```diff
@@ -190,3 +192,23 @@ def test(self, test_folder: str, num_files: int = 20) -> None:
                     os.path.join(test_folder, f"test_mask_ds2_{i}_group{num_mask}.png"),
                     test_sample["label"][1][0, :, :, num_mask],
                 )
+
+
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/dataloading/memseg_dataset.py:196]: dataset token is defined as first token before 1st underscore

File: src/membrain_seg/segmentation/networks/unet.py [hunk 5]
```diff
@@ -102,16 +107,39 @@ def __init__(
...
             res_block=True,
-            # norm_name="INSTANCE",
-            # norm=Norm.INSTANCE,  # I like the instance normalization better than
-            # batchnorm in this case, as we will probably have
-            # only small batch sizes, making BN more noisy
             deep_supervision=True,
             deep_supr_num=2,
         )
-        ignore_dice_loss = IgnoreLabelDiceCELoss(ignore_label=2, reduction="mean")
+
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/networks/unet.py:113]: The following initializes a weighted average between loss functions.
BCE/DICE loss are used by default, Surface-Dice is optionally added

File: src/membrain_seg/segmentation/networks/unet.py [hunk 6]
```diff
@@ -180,16 +210,27 @@ def training_step(
...
         output = self.forward(images)
-        loss = self.loss_function(output, labels)
+        loss = self.loss_function(output, labels, ds_label)
 
         stats_dict = {"train_loss": loss, "train_number": output[0].shape[0]}
         self.training_step_outputs.append(stats_dict)
         self.running_train_acc += (
             masked_accuracy(output[0], labels[0], ignore_label=2.0, threshold_value=0.0)
             * output[0].shape[0]
         )
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/networks/unet.py:222]: also log surface-dice during training

File: src/membrain_seg/segmentation/train.py [hunk 7]
```diff
@@ -52,11 +58,31 @@ def train(
...
+        If True, enables Surface-Dice loss.
+    surf_dice_weight : float, optional
+        Weight for the Surface-Dice loss.
+    surf_dice_tokens : list, optional
+        List of tokens to use for the Surface-Dice loss.
 
     Returns
     -------
     None
     """
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/train.py:71]: This prints a summary of training parameters before each training run.

File: src/membrain_seg/segmentation/training/optim_utils.py [hunk 8]
```diff
@@ -53,7 +53,7 @@ class IgnoreLabelDiceCELoss(_Loss):
     def __init__(
         self,
         ignore_label: int,
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/training/optim_utils.py:55]: Loss functions should be default return non-reduced losses. I.e., for each element in the batch, a single value should be returned.
This way, we can decide whether to apply the respective loss for each element in the batch (depending on the dataset token)

File: src/membrain_seg/segmentation/training/optim_utils.py [hunk 9]
```diff
@@ -144,13 +159,105 @@ def forward(self, inputs: list, targets: list) -> torch.Tensor:
...
-            loss += weight * self.loss_fn(data, target)
+        ds_labels_loop = [ds_labels] * 5
+        for weight, data, target, ds_label in zip(
+            self.weights, inputs, targets, ds_labels_loop
+        ):
+            loss += weight * self.loss_fn(data, target, ds_label)
+        return loss
+
+
+class CombinedLoss(_Loss):
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/training/optim_utils.py:179]: Combined loss function computes the weighted averages of all losses, and considers only specified datasets.
In this way, we can choose exactly which losses to apply for which dataset

File: src/membrain_seg/segmentation/training/surface_dice.py [hunk 10]
```diff
@@ -0,0 +1,455 @@
+"""
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/training/surface_dice.py:1]: Implementation of Surface-Dice functionalities

Comment (LorenzLamm) [reply src/membrain_seg/segmentation/training/surface_dice.py:1]: Soft skeletonization of the segmentation is achieved by iteratively eroding and dilating the membrane, and keeping track of the differences.
Erosion and Dilation can be achieved by min- and max-pooling, respectively. This makes the function differentiable and we can perform backpropagation through it.

File: src/membrain_seg/segmentation/training/surface_dice.py [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+    img1 = soft_open(img, separate_pool=separate_pool)
+    skel = F.relu(img - img1)
+    for _j in range(iter_):
+        img = soft_erode(img)
+        img1 = soft_open(img, separate_pool=separate_pool)
+        delta = F.relu(img - img1)
+        skel = skel + F.relu(delta - skel * delta)
+    return skel
+
+
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/training/surface_dice.py:190]: Defined Gaussian kernel for smoothing binary segmentations before skeletonization.
I didn't find another torch function for this, so I implemented it.
This allows computation of smoothing on GPU without moving stuff between devices.

File: src/membrain_seg/segmentation/training/surface_dice.py [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+    padding = kernel_size // 2
+
+    # Move the kernel to the same device as the segmentation tensor
+    g_kernel = g_kernel.to(seg.device)
+
+    # Apply the Gaussian filter
+    filtered_seg = F.conv3d(seg, g_kernel, padding=padding, groups=seg.shape[1])
+    return filtered_seg
+
+
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/training/surface_dice.py:293]: for binary segmentations, we first perform Gaussian smoothing. Otherwise soft skeletons are discontinuous

File: src/membrain_seg/segmentation/training/surface_dice.py [hunk 13]
```diff
@@ -0,0 +1,455 @@
...
+    - The resulting skeleton is returned as a binary torch.Tensor where 1 indicates the
+        skeleton points.
+    """
+    gt_smooth = (
+        apply_gaussian_filter((gt_seg == 1) * 1.0, kernel_size=15, sigma=2.0) * 1.5
+    )
+    skel_gt = soft_skel(gt_smooth, iter_=iterations)
+    return skel_gt
+
+
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/training/surface_dice.py:336]: Computation of Surface-Dice similar to centerline Dice.

File: src/membrain_seg/segmentation/training/training_param_summary.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
+def print_training_parameters(
```
Comment (LorenzLamm) [inline src/membrain_seg/segmentation/training/training_param_summary.py:1]: I thought it might be useful to print training parameters before each training run. Maybe makes sense to make this optional?

File: tests/membrain_seg/training/test_optim_utils.py
Comment (LorenzLamm) [inline tests/membrain_seg/training/test_optim_utils.py:1]: Adjusted test function to work with new loss definitions.
I should cover more code with the tests I guess :/

File: src/membrain_seg/segmentation/training/surface_dice.py [hunk 15]
```diff
@@ -0,0 +1,455 @@
...
+    """
+    Compute the surface Dice loss with masking for ignore labels.
+
+    The surface Dice loss measures the similarity between the predicted segmentation's
+    skeleton and the ground truth segmentation (and vice versa). Labels annotated with
+    "ignore_label" are ignored.
+
+    Parameters
+    ----------
+    data : torch.Tensor
```
Comment (kevinyamauchi) [inline src/membrain_seg/segmentation/training/surface_dice.py:355]: It looks like the code assumes a certain shape for the data array. It would be nice to write the expected axis order in the docstring (e..g, `[B, C, Z, Y, X]`)

Comment (kevinyamauchi) [reply src/membrain_seg/segmentation/training/surface_dice.py:355]: (same in the class below)

File: src/membrain_seg/segmentation/training/surface_dice.py [hunk 16]
```diff
@@ -0,0 +1,455 @@
...
+    -----
+    This function uses a precomputed dictionary to enhance performance by
+    storing Gaussian kernels. If a kernel with the specified size and standard
+    deviation does not exist in the dictionary, it is created and added. The
+    function assumes the input tensor is a 5D tensor, applies 3D convolution
+    using the Gaussian kernel with padding to maintain spatial dimensions, and
+    it performs the operation separately for each channel of each batch item.
+    """
+    # Create the Gaussian kernel or load it from the dictionary
+    global gaussian_kernel_dict
```
Comment (kevinyamauchi) [inline src/membrain_seg/segmentation/training/surface_dice.py:274]: Is it necessary to use a global here?

File: src/membrain_seg/segmentation/training/surface_dice.py (hunk 11 above)
Comment (kevinyamauchi) [reply src/membrain_seg/segmentation/training/surface_dice.py:190]: Good idea!

File: tests/membrain_seg/training/test_optim_utils.py
Comment (kevinyamauchi) [reply tests/membrain_seg/training/test_optim_utils.py:1]: Thanks for updating. It would be nice to increase the test coverage over time, but it's fine for this PR

File: src/membrain_seg/segmentation/cli/train_cli.py (hunk 2 above)
Comment (kevinyamauchi) [reply src/membrain_seg/segmentation/cli/train_cli.py:89]: ah that's a bummer, but I think it's okay. In the future, it might make sense to add support for a `glob` string or something or directory so that people don't have to write all of the tokens.

File: src/membrain_seg/segmentation/training/optim_utils.py (hunk 8 above)
Comment (kevinyamauchi) [reply src/membrain_seg/segmentation/training/optim_utils.py:55]: Agreed!

File: src/membrain_seg/segmentation/training/optim_utils.py [hunk 17]
```diff
@@ -95,11 +95,26 @@ def forward(self, data: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
+                target[batch_idx].unsqueeze(0),
+                mask[batch_idx].unsqueeze(0),
+            )
 
         # Combine the Dice and Cross Entropy losses
         combined_loss = self.lambda_dice * dice_loss + self.lambda_ce * bce_loss
+        if self.reduction == "mean":
+            combined_loss = combined_loss.mean()
+        elif self.reduction == "sum":
+            combined_loss = combined_loss.sum()
```
Comment (kevinyamauchi) [inline src/membrain_seg/segmentation/training/optim_utils.py:117]: it woudl be nice to add and `else` case that raises an error so that it doesn't fail silently if people make a typo

**Review by kevinyamauchi — APPROVED** [review]
This looks good to me! I have some minor comments below. Based on the conversation on zulip with @alisterburt , it sounds like we are in agreement that this loss function should live in `membrain-seg`. I think you can merge after you address the minor comments. Thanks, @LorenzLamm !

**Comment by LorenzLamm** [comment]
> This looks good to me! I have some minor comments below. Based on the conversation on zulip with @alisterburt , it sounds like we are in agreement that this loss function should live in `membrain-seg`. I think you can merge after you address the minor comments. Thanks, @LorenzLamm !

Cool, thanks a lot for your feedback @kevinyamauchi ! Implemented your suggestions and merging now.
